### ISSUE TITLE
nteract processes alive after closing application window

### ISSUE THREAD

**Issue by jarmokivekas** [nteract/archived-desktop-app#4247]
**Application or Package Used**

nteract desktop application version 0.12.3 installed using .deb installer on ubuntu 18.10

**Describe the bug**

Closing the application window sometimes gives "SHUTDOWN_REPLY_TIMED_OUT: Timeout has occured" notifications before closing. nteract background processes stay alive in the background

**To Reproduce**

1. open notebook and do some work
2. press atl-F4 to close the nteract window (not sure if how the window is closed has any meaning)

- Happens regularly, but I can't intentionally reproduce yet
- after a day of working, some notebooks will close normally, and others give the error before closing.
- running `ps aux | grep nteract` gives a long list of processes even after all windows are closed

**Expected behavior**

- Closing the application window should also kill all background processes.
- Closing the window shuldn't regularly cause shutdown_reply_timed_out errors.

**Screenshots**

This is the error popup just before closing, it comes up after hitting alt-F4 but just berfore the window closes

![screenshot from 2019-03-01 13-16-22](https://user-images.githubusercontent.com/3074453/53661117-44cc0f00-3c68-11e9-8864-07109447d876.png)

There are also multiple nteract icons/menus left in the menubar (Gnome desktop on Ubuntu 18.10).
The icons go away on normal shutdown, but stay in if the shutdown timeout error occurs.

![screenshot from 2019-03-01 20-46-06](https://user-images.githubusercontent.com/3074453/53661275-a68c7900-3c68-11e9-94da-5ad124679310.png)

**Desktop (please complete the following information):**

- OS: Ubuntu 18.10 using Gnome desktop
- Browser: Firefox? I use the nteract desktop application...
- Version: `nteract --version` gives 0.12.3

**Additional context**

Below is verbatim output of `ps aux | grep nteract`. I have no nteract windows open, and expect that this whould not reutrn anything. The second screenshot was taken the same moment. There are the remnant processes of 4 application instances. Two where I opened a notebook, and two there i created new notebooks. Each instance has the  `nteract [filename]` process and one  `--type=zygote` and `--type=gpu-process`.

```
jarmo     3428  0.8  0.8 1436956 66912 tty2    Sl+  10:08   5:24 /opt/nteract/nteract
jarmo     3433  0.0  0.2 252832 18516 tty2     S+   10:08   0:00 /opt/nteract/nteract --type=zygote --no-sandbox
jarmo     3577  0.3  0.4 383172 32068 tty2     Sl+  10:08   2:13 /opt/nteract/nteract --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=801F82A51E592661C285ABB2B18F1CD5
jarmo    10416  0.1  0.7 1415524 59784 ?       Sl   13:16   0:39 /opt/nteract/nteract /home/jarmo/Documents/ELEC-E4210/emission/emission.ipynb
jarmo    10419  0.0  0.2 252832 18504 ?        S    13:16   0:00 /opt/nteract/nteract --type=zygote --no-sandbox
jarmo    10460  0.0  0.3 380636 28360 ?        Sl   13:16   0:02 /opt/nteract/nteract --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=884C45FEBDD668920BBBEBD6BE204ADD
jarmo    11822  0.2  0.8 1416160 64264 tty2    Sl+  13:30   1:10 /opt/nteract/nteract
jarmo    11826  0.0  0.2 252832 18660 tty2     S+   13:30   0:00 /opt/nteract/nteract --type=zygote --no-sandbox
jarmo    11865  0.0  0.3 379480 28048 tty2     Sl+  13:30   0:17 /opt/nteract/nteract --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=8EEBCAB46DD0511191F993E3342C38A0
jarmo    13183  1.5  1.1 1404876 90020 ?       Sl   14:22   6:20 /opt/nteract/nteract /home/jarmo/Documents/ELEC-E4210/emission-mechanics/emission-mechanics-ind.ipynb
jarmo    13186  0.0  0.2 252864 18548 ?        S    14:22   0:00 /opt/nteract/nteract --type=zygote --no-sandbox
jarmo    13224  0.6  0.7 384508 55540 ?        Sl   14:22   2:47 /opt/nteract/nteract --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=71EF779009BFADAA353A105770A1EBAE
jarmo    32621  0.0  0.0  20012   892 pts/1    S+   21:07   0:00 grep --color=auto nteract
```

**Comment by jarmokivekas** [comment]
This might actually be two separate issues. The *shutdown reply timed out*  error doesn't have to happen for the processes to be left running.

**Reproducing reliably**:

1. run `nteract` from command line without arguments
2. wait for application to load
3. press ctrl+F4
4. "Are you sure you want to quit?", click yes.
5. window closes seemingly normallly. `shutdown_reply_timed_out` does **not** happen (no error output at all
6. the icon in the top menu bar never disappears

For every closed application one set of processes like these remain in `ps aux | grep nteract`:

```
jarmo     1557  0.3  1.6 1414500 127132 tty2   Sl+  21:41   0:02 /opt/nteract/nteract
jarmo     1559  0.0  0.5 252832 39936 tty2     S+   21:41   0:00 /opt/nteract/nteract --type=zygote --no-sandbox
jarmo     1603  0.0  0.8 375964 66160 tty2     Sl+  21:41   0:00 /opt/nteract/nteract --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=31DC211893C62621AC94C57D31BA836F
```

**Comment by jarmokivekas** [comment]
Since the v0.12.3 of the desktop application is pretty old, I tried reproducing these bugs on the current master branch (commit 0d10bda47ba5e50), and it's pretty much unchanged, unfortunately.

Steps for reproducing reliably:

1. run `killall nteract` to make sure no earlier instances are running
1. run `yarn app:desktop`
1. the app opens with an empty notebook
1. run `ps aux | grep nteract` in a separate terminal
1. press ctrl-F4 to close the nteract window
1. there is **no** _shutdown reply timed out_ error
1. the window closes as expected, but icons are left in the OS top menu.
1. the `yarn app:desktop` command does not exit (the terminal is still busy)
1. run `ps aux | grep nteract` in separate terminal
1. press ctrl-C to kill `yarn app:desktop`
1. the nteract icon disappears, and all nteract related processes shut down (so everything is as it should).

The process works pretty much the same way on my v0.12.3 installation as on the master, if I use the system-wide `nteract` command instead of running it using `yarn app:desktop`. Process names are a little different, but the behavior is the same.

When nteract is running these processes are active:

```
jarmo     9465  0.0  0.0   2576   736 pts/1    S+   19:43   0:00 /bin/sh -c lerna run dev --stream --scope nteract
jarmo     9466 27.1  1.1 900812 93684 pts/1    Sl+  19:43   0:09 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/node_modules/.bin/lerna run dev --stream --scope nteract
jarmo     9630  1.2  0.4 546580 32560 pts/1    Sl+  19:43   0:00 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/node_modules/.bin/cross-env NODE_ENV=development electron .
jarmo     9637  1.0  0.3 546520 30872 pts/1    Sl+  19:43   0:00 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/.bin/electron .
jarmo     9644 16.8  1.9 1715920 153828 pts/1  Sl+  19:43   0:01 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron .
jarmo     9646  0.3  0.5 252940 40356 pts/1    S+   19:43   0:00 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron --type=zygote --no-sandbox
jarmo     9685  2.5  1.0 386272 83072 pts/1    Sl+  19:43   0:00 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=3CB5372DEA7A728142890B1B40204B79
jarmo     9720 40.0  3.2 1213224 258316 pts/1  Sl+  19:43   0:02 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron --type=renderer --no-sandbox --service-pipe-token=3D144A53C2BBCD86A665632F4D1E02A1 --lang=en-US --app-path=/home/jarmo/Projects/nteract/nteract/applications/desktop --node-integration=true --webview-tag=true --no-sandbox --background-color=#fff --num-raster-threads=2 --enable-main-frame-before-activation --enable-compositor-image-animations --service-request-channel-token=3D144A53C2BBCD86A665632F4D1E02A1 --renderer-client-id=6 --shared-files=v8_context_snapshot_data:100,v8_natives_data:101
jarmo     9768  0.0  0.0  20012   864 pts/2    S+   19:43   0:00 grep --color=auto nteract
```

After closing the window, there is one less active process (the --type=renderer process shuts down)

```
jarmo@jarmo-ThinkPad-T430s:~/Projects/nteract/nteract$ ps aux | grep nteract
jarmo     9465  0.0  0.0   2576   736 pts/1    S+   19:43   0:00 /bin/sh -c lerna run dev --stream --scope nteract
jarmo     9466 21.2  1.1 900812 93684 pts/1    Sl+  19:43   0:09 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/node_modules/.bin/lerna run dev --stream --scope nteract
jarmo     9630  0.5  0.4 546580 32560 pts/1    Sl+  19:43   0:00 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/node_modules/.bin/cross-env NODE_ENV=development electron .
jarmo     9637  0.4  0.3 546520 30872 pts/1    Sl+  19:43   0:00 /usr/local/bin/node /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/.bin/electron .
jarmo     9644  7.8  2.0 1487060 160992 pts/1  Sl+  19:43   0:01 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron .
jarmo     9646  0.1  0.5 252940 40356 pts/1    S+   19:43   0:00 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron --type=zygote --no-sandbox
jarmo     9685  1.1  1.0 375248 78748 pts/1    Sl+  19:43   0:00 /home/jarmo/Projects/nteract/nteract/applications/desktop/node_modules/electron/dist/electron --type=gpu-process --no-sandbox --gpu-preferences=KAAAAAAAAACAAACAAQAAAAAAAAAAAGAAEAAAAAAAAAAAAAAAAAAAAAgAAAAAAAAA --service-request-channel-token=3CB5372DEA7A728142890B1B40204B79
jarmo     9773  0.0  0.0  20012   868 pts/2    S+   19:44   0:00 grep --color=auto nteract
```

After pressing ctrl-C in the terminal, all the nteract related processes shut down. But if you run the installed version of nteract from an application launcher, you don't get to press ctrl-c since there is no terminal.

**Comment by willingc** [comment]
Thanks for the report @jarmokivekas.

**Comment by BenRussert** [comment]
This is fixed in #4332, which we hope to ship next week.

**Comment by jarmokivekas** [comment]
I'm so happy this is fixed now. :heart_eyes:

ctrl + f2, `killall nteract` was pretty much a habit for me... 

Thanks
